main.go

```diff
@@ -279,13 +300,60 @@ func main() {
 			pprof.Handler("block").ServeHTTP(c.Writer, c.Request)
 		})
 	}
-
+	group, gCtx := errgroup.WithContext(context.Background())
+	group.Go(func() error {
+		return HandleSignals(gCtx)
+	})
+	group.Go(func() error {
+		setupPush(gCtx)
+		return nil
+	})
 	// Collector
 	if collectorEnabled {
-		collector()
+		group.Go(func() error {
+			collector(gCtx)
+			return nil
+		})
 	}
 	cleanup()
-	r.Run(":8888")
+	srv := &http.Server{
+		Addr:    ":8888",
+		Handler: r.Handler(),
+	}
+	group.Go(func() error {
+		errorChan := make(chan error, 1)
+		go func() {
+			defer close(errorChan)
+			if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				errorChan <- err
+			}
+		}()
+		select {
+		case <-gCtx.Done():
+			return nil // context cancelled
+		case err := <-errorChan:
+			return err
+		}
+	})
+	if err := group.Wait(); err != nil {
+		switch err.Error() {
+		case "received SIGTERM":
+			log.Println("received SIGTERM shutting down")
+		case "received interrupt":
+			log.Println("received interrupt shutting down")
+		default:
+			log.Fatal(err) // unexpected error
+		}
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := srv.Shutdown(ctx); err != nil {
+		log.Fatal(err) // failure/timeout shutting down the server gracefully
+	}
+	select {
```

Comment:
This can be removed or adapted, as this is only the case when the shutdown context is not completed within 5 seconds.